tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/Continuations/ValueTaskContinuationGenerator`1.NetFx.cs

```diff
@@ -0,0 +1,250 @@
+internal class ValueTaskContinuationGenerator<TIntegration, TTarget, TReturn, TResult> : ContinuationGenerator<TTarget, TReturn, TResult>
+{
+    private static readonly CallbackHandler Resolver;
+
+    static ValueTaskContinuationGenerator()
+    {
+        var result = IntegrationMapper.CreateAsyncEndMethodDelegate(typeof(TIntegration), typeof(TTarget), typeof(TResult));
+        if (result.Method is not null)
+        {
+            if (result.Method.ReturnType == typeof(Task) ||
+                (result.Method.ReturnType.IsGenericType && typeof(Task).IsAssignableFrom(result.Method.ReturnType)))
+            {
+                var asyncContinuation = (AsyncContinuationMethodDelegate)result.Method.CreateDelegate(typeof(AsyncContinuationMethodDelegate));
+                Resolver = new AsyncCallbackHandler(asyncContinuation, result.PreserveContext);
+            }
+            else
+            {
+                var continuation = (ContinuationMethodDelegate)result.Method.CreateDelegate(typeof(ContinuationMethodDelegate));
+                Resolver = new SyncCallbackHandler(continuation, result.PreserveContext);
+            }
+        }
+        else
+        {
+            Resolver = new NoOpCallbackHandler();
+        }
+
+        if (Log.IsEnabled(LogEventLevel.Debug))
+        {
+            Log.Debug(
+                "== {TaskContinuationGenerator} using Resolver: {Resolver}",
+                $"TaskContinuationGenerator<{typeof(TIntegration).FullName}, {typeof(TTarget).FullName}, {typeof(TReturn).FullName}>",
+                Resolver.GetType().FullName);
+        }
+    }
+
+    public override TReturn? SetContinuation(TTarget? instance, TReturn? returnValue, Exception? exception, in CallTargetState state)
+    {
+        return Resolver.ExecuteCallback(instance, returnValue, exception, in state);
+    }
+
+    private class SyncCallbackHandler : CallbackHandler
+    {
+        private readonly ContinuationMethodDelegate _continuation;
+        private readonly bool _preserveContext;
+
+        public SyncCallbackHandler(ContinuationMethodDelegate continuation, bool preserveContext)
+        {
+            _continuation = continuation;
+            _preserveContext = preserveContext;
+        }
+
+        public override TReturn? ExecuteCallback(TTarget? instance, TReturn? returnValue, Exception? exception, in CallTargetState state)
+        {
+            if (exception != null || returnValue == null)
+            {
+                _continuation(instance, default, exception, in state);
+                return returnValue;
+            }
+
+            var previousValueTask = returnValue.DuckCast<IValueTaskOfTResultDuckType>();
+            if (previousValueTask.IsCompletedSuccessfully)
+            {
+                // ok all good, just run synchronously
+                var unwrappedResult = previousValueTask.Result;
+                _continuation(instance, unwrappedResult, exception, in state);
+                return returnValue;
```

Comment:
My only concern is that here we are consuming the result of the value task, and returning the same value task. I guess it's ok, but looks like a double usage.

Reply:
oooh, yeah, good spot, [that's not OK](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1?view=net-9.0):

> - Awaiting the instance multiple times.
> - Calling [AsTask](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1.astask?view=net-9.0) multiple times.
> - Using .Result or .GetAwaiter().GetResult() when the operation hasn't yet completed, or using them multiple times.
> - Using more than one of these techniques to consume the instance.

I think we need to create a new instance here, e.g. 

```suggestion
                var unwrappedResult = previousValueTask.Result;
                _continuation(instance, unwrappedResult, exception, in state);
                return ValueTaskActivator<TReturn, TResult?>.CreateInstance(unwrappedResult);
```

(above doesn't compile yet, need to expose the `CreateInstance` that takes a result)

Reply:
Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/6480/commits/a73c2c7a0c2bedf51cedd347387a170d3abdb274